toolchain/check/eval.cpp

```diff
@@ -1689,6 +1689,76 @@ static auto MakeConstantForBuiltinCall(EvalContext& eval_context,
 
     case SemIR::BuiltinFunctionKind::PrintChar:
     case SemIR::BuiltinFunctionKind::PrintInt:
+    case SemIR::BuiltinFunctionKind::StringAt:{
+      //We need both the string and index to be constant
+      Phase phase = Phase::Concrete;
+      auto str_id = GetConstantValue(eval_context, arg_ids[0], &phase);
+      auto index_id = GetConstantValue(eval_context, arg_ids[1], &phase);
+      //If either isnt constant, we can't evaluate it at compile time
+      if (!str_id.has_value() || !index_id.has_value()) {
+        return MakeNonConstantResult(phase);
+      }
+
+      auto str_struct = eval_context.insts().TryGetAs<SemIR::StructValue>(str_id);
+      if(!str_struct) {
+        return MakeNonConstantResult(phase);
+      }
+
+      auto elements = eval_context.inst_blocks().Get(str_struct->elements_id);
+      if (elements.size() != 2) {
+        return MakeNonConstantResult(phase);
+      }
+
+      auto ptr_const_id = eval_context.constant_values().Get(elements[0]);
+      auto string_literal = eval_context.insts().TryGetAs<SemIR::StringLiteral>(
+        eval_context.constant_values().GetInstId(ptr_const_id));
+      if (!string_literal){
+        return MakeNonConstantResult(phase);
+      }
+      
+      auto string_value = eval_context.sem_ir().string_literal_values().Get(
+        string_literal->string_literal_id);
+      //Get index value
+      auto index_inst = eval_context.insts().TryGetAs<SemIR::IntValue>(index_id);
+      if (!index_inst) {
+        CARBON_CHECK(phase != Phase::Concrete, "Concrete constant integer should be a literal");
+        return MakeNonConstantResult(phase);
+      }
+
+      const auto& index_val = eval_context.ints().Get(index_inst->int_id);
+
+      //Check bounds
+      if (index_val.isNegative()){
+        CARBON_DIAGNOSTIC(ArrayIndexNegative, Error, "index `{0}` is negative",
+                          TypedInt);
+        eval_context.emitter().Emit(
+          eval_context.GetDiagnosticLoc(index_id), ArrayIndexNegative,
+          {.type = eval_context.insts().Get(index_id).type_id(),
+          .value = index_val});
+        return SemIR::ErrorInst::ConstantId;
+      }
+
+      //Check for out of bounds
+      if (index_val.getActiveBits() > 64 ||
+          index_val.getZExtValue() >= string_value.size()) {
+        CARBON_DIAGNOSTIC(ArrayIndexOutOfBounds, Error,
+                          "string index `{0}` is past the end of the string",
+                          TypedInt);
+        eval_context.emitter().Emit(
+            eval_context.GetDiagnosticLoc(index_id), ArrayIndexOutOfBounds,
+            {.type = eval_context.insts().Get(index_id).type_id(),
+             .value = index_val});
+        return SemIR::ErrorInst::ConstantId;
+      }
+
+      auto char_value = static_cast<uint8_t>(string_value[index_val.getZExtValue()]);
+
+      auto int_id = eval_context.ints().Add(llvm::APSInt(llvm::APInt(32, char_value), /*isUnsigned=*/false));
```

Comment:
```suggestion
      auto int_id = eval_context.ints().Add(llvm::APSInt(llvm::APInt(8, char_value), /*isUnsigned=*/false));
```
I would expect an `8` here I think?